File: enigma_terminal/nameforge.py
"""Nameforge: strike a named stamp into the chain.

The player picks a short name and the tool searches for a twelve-word phrase
whose first legacy address carries that name right after the leading `1`. The
phrase is real, the address is real, and the only thing cosmetic about it is
the resemblance to a name.

Two things about this are easy to get wrong, and both would show up as a lie
told to the player rather than as a crash.

**Difficulty is not 58^n.** The character immediately after the `1` is nowhere
near uniform: twenty-two of them land about 4.3% of the time and the other
thirty-four about 0.075%, a spread of roughly sixty. `1Rob` therefore costs
about as much as `1Andy` — one character shorter, one much rarer letter. The
distribution is measured rather than assumed, and lives in
``data/nameforge.json``; every estimate here is computed from it, per name.

**A candidate is expensive.** Each one is a fresh 128 bits of entropy, a BIP-39
phrase, PBKDF2-HMAC-SHA512 over 2048 rounds, five levels of BIP-32 and a
hash160. The Python build manages a couple of hundred a second on one core, so
the estimate this module returns is what the interface must show before the
player commits to waiting — not after.
"""

from __future__ import annotations

import json
import math
import secrets
import time
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path

from .crypto_engine import (
    ExtendedKey,
    entropy_to_mnemonic,
    p2pkh_address,
    to_seed,
)

_DATA_FILE = Path(__file__).resolve().parent.parent / "data" / "nameforge.json"

#: Base58 has no 0, O, I or l — a name containing one can never be struck.
BASE58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

#: The path the stamp is read from. Fixed, because an address only means
#: anything alongside the path that produced it.
PATH = "m/44'/0'/0'/0/0"

MIN_LENGTH = 2
MAX_LENGTH = 6


def _load_distribution() -> dict[str, float]:
    payload = json.loads(_DATA_FILE.read_text(encoding="utf-8"))
    return {str(k): float(v) for k, v in payload["leading"].items()}


LEADING = _load_distribution()


class NameError_(ValueError):
    """Raised when a name cannot be struck, with ``kind`` saying why."""

    def __init__(self, message: str, kind: str) -> None:
        super().__init__(message)
        self.kind = kind


def normalise(name: str) -> str:
    """`andy` and `ANDY` both become `Andy`.

    One spelling per name, so two players asking for the same stamp are asking
    for the same work — and so the estimate shown matches the search run.
    """
    cleaned = name.strip()
    return cleaned[:1].upper() + cleaned[1:].lower()


def validate(name: str) -> str:
    """Normalise and check; raises :class:`NameError_` with a reason."""
    stamp = normalise(name)
    if not (MIN_LENGTH <= len(stamp) <= MAX_LENGTH):
        raise NameError_(
            f"NAME MUST BE {MIN_LENGTH}..{MAX_LENGTH} CHARACTERS. GOT {len(stamp)}.",
            kind="length",
        )
    bad = sorted({c for c in stamp if c not in BASE58})
    if bad:
        raise NameError_(
            "NOT IN THE BASE58 ALPHABET: " + ", ".join(bad)
            + ". THERE IS NO 0, O, I OR l IN AN ADDRESS.",
            kind="alphabet",
        )
    return stamp


def probability(stamp: str) -> float:
    """The chance one candidate carries this stamp.

    The first character uses the measured distribution; the rest are uniform
    over the alphabet, which is what they are that deep into the encoding.
    """
    first = LEADING.get(stamp[0], 0.0)
    return first * (1.0 / len(BASE58)) ** (len(stamp) - 1)


def expected_attempts(stamp: str) -> float:
    """Candidates needed on average. The median is about 0.69 of this."""
    chance = probability(stamp)
    return math.inf if chance <= 0 else 1.0 / chance


#: What the interface calls each band. Cut on expected attempts rather than on
#: length, because length is not what makes a name expensive.
TIERS = (
    (10_000, "COMMON"),
    (500_000, "UNCOMMON"),
    (20_000_000, "RARE"),
    (1_000_000_000, "EPIC"),
    (math.inf, "LEGENDARY"),
)


def tier(stamp: str) -> str:
    attempts = expected_attempts(stamp)
    for ceiling, name in TIERS:
        if attempts < ceiling:
            return name
    return "LEGENDARY"


@dataclass(frozen=True)
class Estimate:
    """What to tell the player before they decide to wait."""

    stamp: str
    tier: str
    attempts: float
    bits: float
    seconds: float

    @property
    def is_long(self) -> bool:
        """Longer than a coffee. The interface warns rather than refuses."""
        return self.seconds > 900


def estimate(stamp: str, rate: float) -> Estimate:
    """Expected work for ``stamp`` at ``rate`` candidates a second."""
    attempts = expected_attempts(stamp)
    return Estimate(
        stamp=stamp,
        tier=tier(stamp),
        attempts=attempts,
        # The entropy actually searched, which is log2 of the odds — not the
        # 128 bits of the phrase, which is a different number entirely.
        bits=math.log2(attempts) if attempts != math.inf else math.inf,
        seconds=attempts / rate if rate > 0 else math.inf,
    )


@dataclass(frozen=True)
class Stamp:
    """A struck stamp: a real phrase, and the address it was struck for."""

    stamp: str
    mnemonic: str
    address: str
    path: str
    attempts: int
    seconds: float
    tier: str
    bits: float

    @property
    def preview(self) -> str:
        """`1Andy` and then the part nobody reads."""
        kept = 1 + len(self.stamp)
        return self.address[:kept] + "•" * (len(self.address) - kept)


def candidate() -> tuple[str, str]:
    """One fresh phrase and the address at :data:`PATH`.

    128 bits from the OS random source — the same procedure a wallet follows,
    which is exactly why the result carries the same warning a wallet does.
    """
    mnemonic = entropy_to_mnemonic(secrets.token_bytes(16))
    node = ExtendedKey.from_seed(to_seed(mnemonic)).derive_path(PATH)
    return mnemonic, p2pkh_address(node.public_key)


def forge(
    name: str,
    *,
    limit: int | None = None,
    deadline: float | None = None,
    on_progress: Callable[[int, str], bool | None] | None = None,
    progress_every: int = 500,
) -> Stamp | None:
    """Search until the stamp is struck, or until told to stop.

    ``on_progress`` is handed the attempt count and the closest address seen so
    far; returning False stops the search. Returns None when it stopped without
    a hit, so the caller can tell "not found yet" from "found".
    """
    stamp = validate(name)
    target = "1" + stamp
    started = time.monotonic()
    attempts = 0
    closest = ""
    closest_score = -1

    while True:
        mnemonic, address = candidate()
        attempts += 1

        if address.startswith(target):
            return Stamp(
                stamp=stamp,
                mnemonic=mnemonic,
                address=address,
                path=PATH,
                attempts=attempts,
                seconds=time.monotonic() - started,
                tier=tier(stamp),
                bits=math.log2(expected_attempts(stamp)),
            )

        # How much of the stamp this one did carry — the near miss the
        # interface shows so a long search has something to report.
        score = 0
        for a, b in zip(address[1:], stamp):
            if a != b:
                break
            score += 1
        if score > closest_score:
            closest_score, closest = score, address

        if limit is not None and attempts >= limit:
            return None
        if deadline is not None and time.monotonic() >= deadline:
            return None
        if on_progress and attempts % progress_every == 0:
            if on_progress(attempts, closest) is False:
                return None


def measure_rate(seconds: float = 1.0) -> float:
    """Candidates a second on this machine, measured rather than assumed."""
    started = time.monotonic()
    count = 0
    while time.monotonic() - started < seconds:
        candidate()
        count += 1
    elapsed = time.monotonic() - started
    return count / elapsed if elapsed > 0 else 0.0


def humanise(seconds: float) -> str:
    """A duration the way a person would say it."""
    if seconds == math.inf:
        return "never"
    if seconds < 1:
        return "under a second"
    if seconds < 90:
        return f"{seconds:.0f} seconds"
    if seconds < 5400:
        return f"{seconds / 60:.0f} minutes"
    if seconds < 172800:
        return f"{seconds / 3600:.1f} hours"
    return f"{seconds / 86400:.1f} days"

File: docs/js/nameforge.js
// Nameforge: strike a named stamp into the chain.
//
// The player picks a short name; the tool searches for a twelve-word phrase
// whose first legacy address carries that name right after the leading `1`.
// The phrase is real and the address is real. The resemblance to a name is the
// only cosmetic part.
//
// Case is a choice. Base58 tells `A` from `a`, so a name can be struck two
// ways: exactly as it was spelled, or in whatever case turns up first. The
// second is cheaper — six letters that each exist in both cases is 32 times
// cheaper — and it strikes names the first cannot, because Base58 keeps `o`
// but not `O`, and `L` but not `l`. Both are honest, and each is priced as
// itself.
//
// Difficulty is not 58^n. The character straight after the `1` is nowhere near
// uniform — twenty-two of them land about 4.3% of the time and the other
// thirty-four about 0.075% — so `1Rob` costs about what `1Andy` costs, one
// character shorter and one much rarer letter. The distribution is measured
// rather than assumed (see data/nameforge.json) and every estimate here is
// computed from it, for the actual name, never from its length.
//
// This file is the arithmetic and the rules. The searching happens in
// nameforge-worker.js, off the main thread, because a candidate costs a
// PBKDF2 and five levels of BIP-32 and there may be millions of them.

import { LEADING } from './leading.js';

export const BASE58 =
  '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

/** The path the stamp is read from. An address only means something with one. */
export const PATH = "m/44'/0'/0'/0/0";

export const MIN_LENGTH = 2;
export const MAX_LENGTH = 6;

/**
 * The address characters that would satisfy this position of a name.
 *
 * Everything else here is built on this. In exact mode a character stands for
 * itself alone; in any-case mode it stands for both of its cases — but only
 * those Base58 has, and Base58 is not symmetric: it keeps `o` but not `O`,
 * `i` but not `I`, `L` but not `l`. Twenty-three letters have two forms,
 * three have one, and `0` has none in either mode.
 */
export function variants(char, anyCase = false) {
  if (!anyCase) return BASE58.includes(char) ? char : '';
  const both = [char.toLowerCase(), char.toUpperCase()];
  return [...new Set(both)].filter((c) => BASE58.includes(c)).join('');
}

/**
 * Trim, and nothing else. This used to force `andy` and `ANDY` both to `Andy`,
 * one spelling per name — a stand-in for a decision the player was never
 * offered. They choose the case themselves now.
 */
export const normalise = (name) => String(name).trim();

/**
 * Trim and check. Returns `{ stamp }` or `{ error, kind }` rather than
 * throwing: every caller here is painting a field as the player types.
 */
export function validate(name, anyCase = false) {
  const stamp = normalise(name);
  if (stamp.length < MIN_LENGTH || stamp.length > MAX_LENGTH) {
    return { error: 'length', length: stamp.length };
  }
  const bad = [...new Set([...stamp])].filter((c) => !variants(c, anyCase));
  if (bad.length) return { error: 'alphabet', bad };
  return { stamp };
}

/** Does this address carry the stamp straight after its leading `1`? */
export function matches(address, stamp, anyCase = false) {
  const carried = address.slice(1, 1 + stamp.length);
  return anyCase
    ? carried.toLowerCase() === stamp.toLowerCase()
    : carried === stamp;
}

/**
 * The chance one candidate carries this stamp.
 *
 * Each position contributes the share of addresses whose character there is
 * one this name accepts, so the same arithmetic prices both modes: any-case
 * simply accepts more characters per position.
 */
export function probability(stamp, anyCase = false) {
  let chance = 0;
  for (const c of variants(stamp[0], anyCase)) chance += LEADING[c] ?? 0;
  for (const char of stamp.slice(1))
    chance *= variants(char, anyCase).length / BASE58.length;
  return chance;
}

/** Candidates needed on average. The median is about 0.69 of this. */
export function expectedAttempts(stamp, anyCase = false) {
  const chance = probability(stamp, anyCase);
  return chance > 0 ? 1 / chance : Number.POSITIVE_INFINITY;
}

//: Bands cut on expected attempts, not on length — length is not what makes a
//: name expensive.
export const TIERS = [
  [10_000, 'COMMON'],
  [500_000, 'UNCOMMON'],
  [20_000_000, 'RARE'],
  [1_000_000_000, 'EPIC'],
  [Number.POSITIVE_INFINITY, 'LEGENDARY'],
];

export const tier = (stamp, anyCase = false) => {
  const attempts = expectedAttempts(stamp, anyCase);
  return (TIERS.find(([ceiling]) => attempts < ceiling) || TIERS.at(-1))[1];
};

/** What to tell the player before they decide to wait. */
export function estimate(stamp, rate, anyCase = false) {
  const attempts = expectedAttempts(stamp, anyCase);
  return {
    stamp,
    anyCase,
    tier: tier(stamp, anyCase),
    attempts,
    // The entropy actually searched — log2 of the odds. Not the 128 bits of
    // the phrase, which is a different number and would flatter the result.
    bits: Number.isFinite(attempts)
      ? Math.log2(attempts)
      : Number.POSITIVE_INFINITY,
    seconds: rate > 0 ? attempts / rate : Number.POSITIVE_INFINITY,
  };
}

/** `1Andy` and then the part nobody reads. */
export const preview = (address, stamp) => {
  const kept = 1 + stamp.length;
  return (
    address.slice(0, kept) + '•'.repeat(Math.max(0, address.length - kept))
  );
};

/** A duration the way a person would say it. */
export function humanise(seconds) {
  if (!Number.isFinite(seconds)) return 'never';
  if (seconds < 1) return 'under a second';
  if (seconds < 90) return `${Math.round(seconds)} seconds`;
  if (seconds < 5400) return `${Math.round(seconds / 60)} minutes`;
  if (seconds < 172800) return `${(seconds / 3600).toFixed(1)} hours`;
  return `${(seconds / 86400).toFixed(1)} days`;
}

/** How many workers to run: busy enough to be fast, not so many it locks up. */
export const workerCount = () =>
  Math.max(1, Math.min(8, (navigator.hardwareConcurrency || 4) - 1));
